workers/unity/Packages/io.improbable.worker.sdk.mobile/CopyLibcpp.cs

```diff
@@ -0,0 +1,40 @@
+using System.Collections.Generic;
+using System.IO;
+using UnityEditor;
+using UnityEditor.Android;
+
+namespace Improbable.Gdk.Mobile
+{
+    // It seems that since 14.4.0, the Worker SDK no longer links in libc++_shared.so and Gradle doesn't
+    // properly link it in either.
+    // So we just manually copy it across.
+    // ¯\_(ツ)_/¯
+    internal class CopyLibcpp : IPostGenerateGradleAndroidProject
+    {
+        public int callbackOrder { get; }
+
+        private static readonly string NdkLibRoot = Path.Combine(EditorPrefs.GetString("AndroidNdkRootR16b"), "sources",
+            "cxx-stl", "llvm-libc++", "libs");
+
+        public void OnPostGenerateGradleAndroidProject(string path)
+        {
+            foreach (var arch in GetTargetArchitectures(path))
+            {
+                CopyLibcppSo(arch);
+            }
+        }
+
+        private static IEnumerable<DirectoryInfo> GetTargetArchitectures(string gradleRootPath)
+        {
+            var jniLibsPath = Path.Combine(gradleRootPath, "src", "main", "jniLibs");
+            var dirInfo = new DirectoryInfo(jniLibsPath);
+            return dirInfo.EnumerateDirectories();
+        }
+
+        private static void CopyLibcppSo(DirectoryInfo architectureFolder)
+        {
+            var architecture = architectureFolder.Name;
+            File.Copy(Path.Combine(NdkLibRoot, architecture, "libc++_shared.so"), Path.Combine(architectureFolder.FullName, "libc++_shared.so"));
```

Comment:
should we store `libc++_shared.so` in a field as we use it twice here

Reply:
+1

Reply:
Could do - although given that we won't touch this code much I'm okay with it being like this. Also the locality is nice

Reply:
![](https://media0.giphy.com/media/h20lkWNZWpPAgx7vW4/giphy.gif)